RabbitMQ Healthcheck Contribution

ipeternella opened the pull request:
This pull request contains the following implementations:

-  A class to perform the healthcheck of a RabbitMQ service;
- Tests for the RabbitMQ Healthchecker;
- An updated README.rst with the instructions on how to configure the RabbitMQ healthchecker.

ipeternella committed:
implemented rabbitmq healthcheck and unit tests

ipeternella committed:
added new tests for rabbitmq. Updated README.rst

codingjoe commented on .gitignore:
```diff
@@ -102,3 +102,8 @@ ENV/
 
 # pytest
 .pytest_cache/
+
```
Please add editor specific ignores to you global git ignore. Not to a repository.

ipeternella replied:
Hello there @codingjoe! I have removed these files to a global git ignore. Sorry for this mistake.

codingjoe commented on health_check/contrib/rabbitmq/apps.py:
```diff
@@ -0,0 +1,11 @@
+
+    def ready(self):
+        from .backends import RabbitMQHealthCheck
+        plugin_dir.register(RabbitMQHealthCheck)
```
missing newline

ipeternella replied:
Done!

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+class RabbitMQHealthCheck(BaseHealthCheckBackend):
+    """
+    Healthchecker for RabbitMQ.
+
```
Please drop trailing blank line

ipeternella replied:
Done!

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+    def __get_broker_url(self):
+        """
+        Gets the broker url based on django.confg.settings dictionary.
+
```
Please drop trailing blank line

ipeternella replied:
I have removed this method due to the mentioned complexity of the configuration.

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+        """
+        broker_url = getattr(settings, "BROKER_URL", None)
+
+        if broker_url is None:
```
I have the feeling this drives complexity. I'd rather have only one approach on how to configure this correctly.

ipeternella replied:
I have removed this method due to the mentioned complexity. Right now, the RabbitMQ healthchecker only looks for a BROKER_URL variable on django settings dict.

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+                broker_vhost,
+            )
+
+        return broker_url
```
This can return `None`

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+
+    """
+
+    def __get_broker_url(self):
```
Why use a dunder method here?

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+
+            logger.info("Got broker_url: %s. Creating a connection to the broker...", broker_url)
+
+            conn = pika.BlockingConnection(pika.URLParameters(broker_url))
```
I am not so sure about pika. Most people will use celery and it's `amqplib` fork.

ipeternella replied:
I have removed pika from the dependencies. Instead of it, I use Kombu which is a dependency of the Celery package and therefore comes together with Celery! Kombu uses pyamqp by default according to its docs http://docs.celeryproject.org/projects/kombu/en/latest/userguide/connections.html

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,86 @@
+            logger.info("RabbitMQ is healthy.")
+
+        except ConnectionClosed as e:
+            self.add_error(ServiceUnavailable("Unable to connect to RabbitMQ: Connection refused."), e)
```
You can also just use `raise ServiceUnavailable("your message") from e`

codingjoe commented on requirements-dev.txt:
```diff
@@ -9,3 +9,4 @@ pydocstyle
 pep8-naming
 pytest
 pytest-django
+pika
```
yet another dependency...

ipeternella replied:
Removed pika dependency.

codingjoe commented on setup.cfg:
```diff
@@ -35,6 +35,9 @@ norecursedirs=venv env .eggs
 addopts = --tb=short -rxs
 
+[pep8]
+max-line-length = 119
```
Unrelated change.

ipeternella replied:
Removed this unnecessary change.

codingjoe commented on tests/test_rabbitmq.py:
```diff
@@ -0,0 +1,96 @@
+
+
+if __name__ == "__main__":
+    unittest.main()
```
Kudos for writing tests. I really appreciate it. But we use pytest as a test runner.

ipeternella replied:
Removed from the tests.

codingjoe reviewed (commented):
Hi @IgooorGP great work! Really appreciate the effort. Especially about including tests!

I do have some comments. Please also make sure to add documentation.

ipeternella committed:
Removed pika dependency by using Kombu connection. Changed gitignore and README.

ipeternella commented:
Hello @codingjoe!

Thanks for the code review. I have implemented the requested changes. Please, could you take another look to see if it is acceptable now? Should you have any comments or new requests, please do so.

Thanks a lot!

codingjoe committed:
Update .gitignore

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,51 @@
+        broker_url = getattr(settings, "BROKER_URL", None)
+
+        try:
+            logger.info("Got %s as the broker_url. Connecting to rabbit...")
```
Argument for `%s` is missing

codingjoe replied:
please use debug level here

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,51 @@
+            conn = Connection(broker_url)
+
+            # opens and closes a channel to rabbitmq to make sure the service is up
+            logger.info("Openning a channel to RabbitMQ...")
```
same here, debug please

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,51 @@
+            conn.channel()
+
+            logger.info("Closing the channel to RabbitMQ...")
+            conn.close()
```
is there a context manager you could use? Even if something fails the connection should be closed

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,51 @@
+
+        broker_url = getattr(settings, "BROKER_URL", None)
+
+        try:
```
There is too much happening in the try block. Please be more narrow, where you expect the exception to occour.

codingjoe commented on health_check/contrib/rabbitmq/backends.py:
```diff
@@ -0,0 +1,51 @@
+        Checks if RabbitMQ service is up by opening and closing
+        a broker channel.
+        """
+        logger.info("Checking for a broker_url on django settings...")
```
debug please

codingjoe commented on tests/test_rabbitmq.py:
```diff
@@ -0,0 +1,88 @@
+from health_check.contrib.rabbitmq.backends import RabbitMQHealthCheck
+
+
+class TestRabbitMQHealthCheck(unittest.TestCase):
```
Again, we don't use `UnitTest` but `pytest`. Please just use a call and switch the assert to simple `assert` instead of `self.assertEqual` and co

codingjoe reviewed (changes requested):
Almost there! Just some small changes and this is ready to roll. Thanks 👍

ipeternella committed:
Changed kombu connection to be used as context. Simplified healthcheck test with the connect method.]

ipeternella committed:
Merge branch 'master' of https://github.com/IgooorGP/django-health-check

ipeternella commented:
Hey there @codingjoe ! Thanks again for the review. I have changed the code to use the kombu connection as a context to release opened resources. Also, I decided to use the connect method of the kombu library for the healthcheck because it is easier to understand, produces the same result as before and solves the problem with too many statements in the try/except blocks. Oh, of course, I have changed the tests as well. Please, could you review the code again to make sure everything's all right?

Thanks a lot!

codingjoe committed:
improve try except block

codingjoe committed:
fix doc stings

codingjoe committed:
fix doc strings

codingjoe committed:
fix doc strings

codingjoe committed:
fix doc strings

codingjoe commented:
Released in [3.7.0](https://github.com/KristianOellegaard/django-health-check/releases/tag/3.7.0)
